### ISSUE TITLE
PV's sometimes orphaned and left attached, NodeUnpublishVolume never get's called

### ISSUE THREAD

**Issue by jukie** [kubernetes-sigs/aws-ebs-csi-driver#1798]
/kind bug

**What happened?**

volumeAttachments aren't always properly detached and deleted which leads to PV's with reclaim policy set to `Delete` left sitting at `Released` state with no way to reconcile besides manual intervention. Manually deleting the volumeAttachment resource leads to proper reconciliation through the controller.

Example log from controller:
```
"E1024 23:40:59.094402       1 controller.go:1007] error syncing volume \"pvc-604d4139-974b-4425-a2b1-42c1eb2ae1e4\": persistentvolume pvc-604d4139-974b-4425-a2b1-42c1eb2ae1e4 is still attached to node ip-10-99-9-171.us-west-2.compute.internal",
```
The pod and claim has already been cleaned up but the volume is still attached to the instance. The attacher sidecar doesn't appear to be attempting to detach it and nothing attempts to delete the associated `volumeAttachment` resources.

**What you expected to happen?**
Once the pods are deleted the PV, PVC, and volumeAttachment should all be deleted as well.

**How to reproduce it (as minimally and precisely as possible)?**
Create a large number of pods (e.g. 15x every 2min over 30min) with the example below and after 2hrs you'll likely have some leaked PV's stuck in `Released`
```
apiVersion: v1
kind: Pod
metadata:
  name: ephemeral-test
spec:
  containers:
  - image: alpine
    imagePullPolicy: IfNotPresent
    command:
    - /bin/sh
    - -c
    - sleep 1800
    name: ephemeral-test
    volumeMounts:
    - mountPath: /app
      name: scratch-volume
  volumes:
    - name: scratch-volume
      ephemeral:
        volumeClaimTemplate:
          spec:
            accessModes: [ "ReadWriteOnce" ]
            storageClassName: standard-ssd-encrypted
            resources:
              requests:
                storage: 1Gi
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
  name: standard-ssd-encrypted
parameters:
  encrypted: "true"
  fsType: ext4
  type: gp3
provisioner: ebs.csi.aws.com
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
allowVolumeExpansion: true
```
**Anything else we need to know?**:

We're using generic ephemeral volumes backed by EBS to run various jobs which execute as naked pods behind a custom controller. At peak times,in a span of ~2-3min  this can approach upwards of 10-20 executions with 1 ephemeral EBS volume each.

**Environment**
- Kubernetes version (use `kubectl version`): v1.25.12-eks-2d98532
- Driver version: v1.24.0

We're using the helm chart with the following extra args for sidecars:
Attacher args:
```
        - --worker-threads=300
        - --kube-api-qps=300
        - --kube-api-burst=450
        - --timeout=3m
        - --resync=20s
        - --http-endpoint=:8080
```
Provisioner args:
```
        - --worker-threads=500
        - --kube-api-qps=300
        - --kube-api-burst=450
        - --feature-gates=HonorPVReclaimPolicy=true
        - --timeout=3m
        - --http-endpoint=:8081
```

Why these specific values? We suspect we're hitting a race condition between Detach and Delete operations or breaching the default qps thresholds. How can we verify this?

**Comment by jukie** [comment]
Below is the describe output for the PV:
```
$ kubectl describe pv pvc-1031a27d-848e-4efe-a602-d3dbb556c73d
Name:              pvc-1031a27d-848e-4efe-a602-d3dbb556c73d
Labels:            <none>
Annotations:       pv.kubernetes.io/provisioned-by: ebs.csi.aws.com
                   volume.kubernetes.io/provisioner-deletion-secret-name: 
                   volume.kubernetes.io/provisioner-deletion-secret-namespace: 
Finalizers:        [external-provisioner.volume.kubernetes.io/finalizer kubernetes.io/pv-protection external-attacher/ebs-csi-aws-com]
StorageClass:      standard-ssd-encrypted
Status:            Released
Claim:             redacted-namespace/claim-name
Reclaim Policy:    Delete
Access Modes:      RWO
VolumeMode:        Filesystem
Capacity:          12Gi
Node Affinity:     
  Required Terms:  
    Term 0:        topology.ebs.csi.aws.com/zone in [us-west-2a]
Message:           
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            ebs.csi.aws.com
    FSType:            ext4
    VolumeHandle:      vol-redacted
    ReadOnly:          false
    VolumeAttributes:      storage.kubernetes.io/csiProvisionerIdentity=1698182493322-7313-ebs.csi.aws.com
Events:
  Type     Reason              Age                  From                                                                                      Message
  ----     ------              ----                 ----                                                                                      -------
  Warning  VolumeFailedDelete  106s (x32 over 91m)  ebs.csi.aws.com_ebs-csi-controller-6956b5ffbb-545mf_34101bda-969f-4f95-a512-af95c6fe8aec  persistentvolume pvc-1031a27d-848e-4efe-a602-d3dbb556c73d is still attached to node ip-10-99-9-171.us-west-2.compute.internal
```

And the volumeAttachment:
```
Name:         csi-64bdc8904b085d3a8d53dde7ece69d22f3f2ed21a44c918640fb1d58d9b9bcaf
Namespace:    
Labels:       <none>
Annotations:  csi.alpha.kubernetes.io/node-id: i-05804faeb49354b90
API Version:  storage.k8s.io/v1
Kind:         VolumeAttachment
Metadata:
  Creation Timestamp:  2023-10-24T22:57:44Z
  Finalizers:
    external-attacher/ebs-csi-aws-com
  Resource Version:  682828326
  UID:               b6593d92-6109-429f-903c-1a534decbe76
Spec:
  Attacher:   ebs.csi.aws.com
  Node Name:  ip-10-99-9-171.us-west-2.compute.internal
  Source:
    Persistent Volume Name:  pvc-1031a27d-848e-4efe-a602-d3dbb556c73d
Status:
  Attached:  true
  Attachment Metadata:
    Device Path:  /dev/xvdaa
Events:           <none>
```

The associated claim has already been deleted

**Comment by jukie** [comment]
I've tracked this to occurring when a daemonset pod from either aws-node or ebs-csi-node is restarted due to failing liveness checks for seemingly inexplicable reasons but even if we can prevent the pods from crashing/restarting the driver should tolerate these scenarios and cleanup PV's afterwards vs abandoning them.

**Comment by ConnorJC3** [comment]
I've put this on my backlog to look into (in case this is a bug in Kubernetes itself that we need to report there), but I would recommend reaching out to AWS support if you haven't already

Creation/deletion of `VolumeAttachment` objects is the `kube-controller-manager`'s responsibility; i.e. for an EKS cluster, EKS's responsibility.

**Comment by jukie** [comment]
Thanks, will continue digging there a big.

**Comment by jukie** [comment]
I've confirmed that Detach is never called from kube-controller-manager in these cases and will open a support case with AWS. Closing for now but might reply here once I have more details for guidance on where to open followup issues.

**Comment by jukie** [comment]
@ConnorJC3 reopening, it appears that NodeUnpublishVolume is what never get's called in the daemonset pod. I'm trying to understand the scenarios or edge cases which could lead to this.

**Comment by jukie** [comment]
Last update, this appears to be a known issue that is fixed in k8s v1.27

Summary from a shared slide deck:
>https://github.com/kubernetes/kubernetes/pull/113145
> The kubelet is responsible for notifying the pod worker of both normal updates (spec or status changes), runtime status updates, and periodic resync.  It uses the pod manager as that source.  However, because the pod manager doesn’t know about orphaned pods, it can’t drive the pod worker state machine.  This led to pods not terminating in all cases once the source pod was removed (most noticed for static pods which are force deleted on update by design).  This leaks resources - generally CPU, memory, and volumes, but increasingly will include allocated CPUs via cpu_manager, block devices managed by other components, and GPUs.
> 
> The HandlePodCleanups method is the “resync” mechanism in the kubelet - it is responsible for verifying the desired and actual state match between pod manager, pod worker, and container runtime.  It wasn’t doing that for all cases, so we changed it to be much more thorough.  It is now responsible for calculating orphaned pods, desired pods, actual pods, and taking the necessary actions to ensure the pod worker is in sync with the desired state and the runtime state.  Since this method is called every 2 seconds it will keep notifying pod worker of any orphaned pods until they reach the natural completion of their lifecycle.
> 
> Because we need to take action on orphaned pods, the pod worker now tracks the most recent pod state it has observed, which means it is now the source of truth for “what pods should actually be running right now” in the Kubelet.
